Find invalid entries in homs

slwu89 opened the pull request:
Address #57 by defining a new function `undefined_subparts` which, given a hom, finds for what parts in the domain it is undefined. The most common use is after calling `rem_parts!` on an object, homs with that object as a codomain will have undefined values and we'd like to find them.

This function is only defined for homs whose domain object is `<: DenseParts`, because when `rem_parts!` is used on `<: MarkAsDeleted`, only out-homs are deleted, so given an in-hom it won't be possible to recover the indices of parts "incident" to the undefined value. Example in code below:

```julia
julia> MySch = BasicSchema(
           [:A,:B], [(:f,:A,:B)], [], []
       )
BasicSchema{Symbol}([:A, :B], [(:f, :A, :B)], Symbol[], Tuple{Symbol, Symbol, Symbol}[])

julia> @acset_type MyDataNoIdxBit(MySch, part_type=BitSetParts)
MyDataNoIdxBit

julia> data = @acset MyDataNoIdxBit begin
           A = 5
           B = 5
           f = [1,2,3,4,5]
       end
MyDataNoIdxBit {A:5, B:5}
┌───┬───┐
│ A │ f │
├───┼───┤
│ 1 │ 1 │
│ 2 │ 2 │
│ 3 │ 3 │
│ 4 │ 4 │
│ 5 │ 5 │
└───┴───┘


julia> rem_parts!(data, :B, 1:3)


julia> data
MyDataNoIdxBit {A:5, B:2}
┌───┬───┐
│ A │ f │
├───┼───┤
│ 1 │ 1 │
│ 2 │ 2 │
│ 3 │ 3 │
│ 4 │ 4 │
│ 5 │ 5 │
└───┴───┘


julia> data.subparts[:f]
ACSets.ColumnImplementations.DenseFinColumn{Vector{Int64}}(ACSets.Mappings.VecMap(1 => 1, 2 => 2, 3 => 3, 4 => 4, 5 => 5), ACSets.PreimageCaches.TrivialCache{Int64, Int64}())
```

kris-brown commented on src/DenseACSets.jl:
```diff
@@ -644,6 +644,26 @@ end
+end
+
+function _undefined_subparts(acs::SimpleACSet, f::Symbol, ::MarkAsDeleted)
+  error("undefined_subparts only defined for homs with DenseParts domains")
```
Would it make sense for `undefined_subparts` to not error here but instead return the (undeleted) parts of the domain which are pointing at a part that has been marked as deleted in the codomain?

slwu89 replied:
Good point! Yes, I'll implement that.

kris-brown commented on src/DenseACSets.jl:
```diff
@@ -644,6 +644,26 @@ end
+  _undefined_subparts(acs, f, acs.parts[dom(s,f)])
+end
+
+function ACSetInterface.undefined_subparts(acs::DynamicACSet, f::Symbol)
```
It might be unnecessary to make a different method for DynamicACSets. You can get the schema for any ACSet by calling `acset_schema(acs)`.

kris-brown commented on test/ACSets.jl:
```diff
@@ -73,7 +73,18 @@ for dds_maker in dds_makers
 
   # Deletion.
+  @test_throws ErrorException undefined_subparts(dds, :X)
+  if dds.parts[dom(acset_schema(dds),:Φ)] isa DenseParts
```
This might be a good time to implement a function which gets the parts type from an ACSet, which is known at the type level. E.g. `parts_type(::ACSet{PT}) where PT = PT`. 

Then this test could become `if parts_type(dds) <: DenseParts `

slwu89 replied:
Great suggestion.

kris-brown replied:
unfortunately now this function isn't being tested - maybe you can throw a line in somewhere like `@test typeof(dds.parts[:X]) isa parts_type(dds)`

slwu89 replied:
whoops.

slwu89 commented:
Thanks @kris-brown for the review! I've addressed your comments, and now the `undefined_subparts` method works for acsets with either `<:DenseParts` or `<:MarkAsDeleted` implementation of parts ids.

kris-brown commented on src/DenseACSets.jl:
```diff
@@ -644,6 +652,21 @@ end
+function ACSetInterface.undefined_subparts(acs::SimpleACSet, f::Symbol)
+  s = acset_schema(acs)
+  f ∈ homs(s; just_names=true) || error("$f is not a hom")
+  _undefined_subparts(acs, f, parts_type(acs))
```
I don't think `parts_type` is actually needed to be called for dispatch since you can do 
```
function undefined_subparts(acs::SimpleACSet{<:DenseParts}, f::Symbol) ... end 

function undefined_subparts(acs::SimpleACSet{<:MarkAsDeleted}, f::Symbol) ... end 
``` 

If there's not much shared code between the two implementations you don't need a top level `undefined_parts` that calls `_undefined_parts`. (In particular, I think checking that `f` is in `homs(s)` is implicitly done by indexing `acs.subparts[f]` in each of the implementations.)

kris-brown commented on src/DenseACSets.jl:
```diff
@@ -644,6 +652,15 @@ end
+end
+
+function ACSetInterface.undefined_subparts(acs::SimpleACSet{<:MarkAsDeleted}, f::Symbol)
+  codom_ids = codom_parts(acs,f)
```
oh never mind, seems like MarkAsDeleted ACSets naturally store things in a `BitSet`

slwu89 replied:
good to keep in mind in case other subtypes of `MarkAsDeleted` are used in the future!

kris-brown commented:
Thanks @slwu89, this looks good to merge in once you add an extra test line for `parts_type`!

slwu89 commented:
Thanks @kris-brown! Tests for `parts_type` added in [6fe42eb](https://github.com/AlgebraicJulia/ACSets.jl/pull/62/commits/6fe42eb6558834e4c78324b05a5005b6be8d831e)
